File: example/python/yield_curve.py
"""Bootstrap a yield curve from deposits and swaps, then query it.

`PiecewiseYieldCurve` takes a bundle of rate helpers (each pinning one market
quote) and solves for the discount factors that reprice all of them at once.
The curve is lazy: nothing is solved until the first query.

The deposit and swap quotes are transcribed from QuantLib's
`piecewiseyieldcurve.cpp`. The pedagogical payoff and the self-check are the
same thing: a FRESH Euribor index laid on the bootstrapped curve must forecast
back each deposit's own input rate. If it does, the bootstrap is consistent.

Run it with:

    python example/python/yield_curve.py
"""

# plugins
# itofin library
from itofin import Settings
from itofin.indexes import Euribor
from itofin.quotes import SimpleQuote
from itofin.termstructures import DepositRateHelper, PiecewiseYieldCurve, SwapRateHelper
from itofin.time import BusinessDayConvention, Calendar, Date, DayCounter, Frequency
from itofin.time import Period as P

# (n, unit, rate-in-percent) short-end deposits.
DEPOSIT_DATA = [
    (1, "Weeks", 4.559),
    (1, "Months", 4.581),
    (3, "Months", 4.557),
    (6, "Months", 4.496),
    (9, "Months", 4.490),
]

# (years, rate-in-percent) par swap rates for the long end.
SWAP_DATA = [
    (1, 4.54),
    (2, 4.63),
    (3, 4.75),
    (5, 4.99),
    (10, 5.47),
    (20, 5.89),
    (30, 5.96),
]


def main() -> None:
    settings = Settings()
    calendar = Calendar.target()
    today = calendar.adjust(Date(15, 6, 2026), BusinessDayConvention.Following)
    settings.set_evaluation_date(today)

    # Settlement is spot: today advanced two business days.
    settlement = calendar.advance(today, 2, "Days", BusinessDayConvention.Following, False)

    # Deposit helpers: each quotes a Euribor rate over an (as yet unlinked)
    # forwarding index. The SimpleQuote holds the market rate.
    deposits = []
    for n, unit, rate in DEPOSIT_DATA:
        index = Euribor(P(n, unit), None, settings)
        deposits.append(DepositRateHelper(SimpleQuote(rate / 100.0), index))

    # Swap helpers: a par swap rate, an annual fixed leg on 30/360 bond basis,
    # floating off 6M Euribor.
    swaps = []
    for n, rate in SWAP_DATA:
        euribor6m = Euribor(P(6, "Months"), None, settings)
        swaps.append(
            SwapRateHelper(
                SimpleQuote(rate / 100.0),
                P(n, "Years"),
                calendar,
                Frequency.Annual,
                BusinessDayConvention.Unadjusted,
                DayCounter.thirty360_bond_basis(),
                euribor6m,
            )
        )

    # Log-linear-on-discount-factors interpolation (the market default).
    curve = PiecewiseYieldCurve(settlement, deposits + swaps, DayCounter.actual360(), "LogLinear")

    # First query forces the (lazy) bootstrap. Query strictly inside the curve
    # span; t=0 and points past the last pillar are range-rejected.
    print("Discount factors and zero rates off the bootstrapped curve:")
    for t in [0.5, 1.0, 2.0, 5.0, 10.0, 20.0]:
        df = curve.discount(t)
        zero = curve.zero_rate(t)
        print(f"  t={t:5.1f}y   discount={df:.6f}   zero={zero * 100:6.3f}%")

    # Consistency check: a fresh index on the solved curve reforecasts each
    # deposit's own input rate. This is the bootstrap contract made visible.
    print("\nDeposit reprice check (fresh index on the curve vs input quote):")
    for n, unit, rate in DEPOSIT_DATA:
        index = Euribor(P(n, unit), curve, settings)
        forecast = index.fixing(today, False)
        print(
            f"  {n:2d} {unit:<7} forecast={forecast * 100:6.3f}%   "
            f"input={rate:6.3f}%   |diff|={abs(forecast - rate / 100.0):.2e}"
        )


if __name__ == "__main__":
    main()

File: crates/libitofin/examples/yield_curve.rs
//! Bootstrap a `PiecewiseYieldCurve` from deposit + swap rate helpers, then
//! query discount factors and zero rates off the solved curve.
//!
//! The market strip is transcribed from QuantLib's piecewiseyieldcurve.cpp.
//! D5: the evaluation date is set explicitly on an owned `Settings` (an unset
//! date is an Err, not a system-clock fallback).

use libitofin::handle::Handle;
use libitofin::indexes::ibor::euribor::Euribor;
use libitofin::interestrate::Compounding;
use libitofin::math::interpolations::loglinear::LogLinear;
use libitofin::quotes::{Quote, SimpleQuote};
use libitofin::settings::Settings;
use libitofin::shared::{Shared, shared};
use libitofin::termstructures::TermStructure;
use libitofin::termstructures::bootstraphelper::RateHelper;
use libitofin::termstructures::bootstraptraits::Discount;
use libitofin::termstructures::yields::{DepositRateHelper, PiecewiseYieldCurve, SwapRateHelper};
use libitofin::termstructures::yieldtermstructure::YieldTermStructure;
use libitofin::time::businessdayconvention::BusinessDayConvention;
use libitofin::time::calendars::target::Target;
use libitofin::time::date::{Date, Month};
use libitofin::time::daycounters::actual360::Actual360;
use libitofin::time::daycounters::thirty360::{Convention, Thirty360};
use libitofin::time::frequency::Frequency;
use libitofin::time::period::Period;
use libitofin::time::timeunit::TimeUnit;
use libitofin::types::Rate;

// Market strip transcribed from QuantLib's piecewiseyieldcurve.cpp
// (deposits :364-378, swaps :379-403). (n, unit, rate-in-percent).
const DEPOSIT_DATA: [(i32, TimeUnit, f64); 6] = [
    (1, TimeUnit::Weeks, 4.559),
    (1, TimeUnit::Months, 4.581),
    (2, TimeUnit::Months, 4.573),
    (3, TimeUnit::Months, 4.557),
    (6, TimeUnit::Months, 4.496),
    (9, TimeUnit::Months, 4.490),
];

const SWAP_DATA: [(i32, TimeUnit, f64); 15] = [
    (1, TimeUnit::Years, 4.54),
    (2, TimeUnit::Years, 4.63),
    (3, TimeUnit::Years, 4.75),
    (4, TimeUnit::Years, 4.86),
    (5, TimeUnit::Years, 4.99),
    (6, TimeUnit::Years, 5.11),
    (7, TimeUnit::Years, 5.23),
    (8, TimeUnit::Years, 5.33),
    (9, TimeUnit::Years, 5.41),
    (10, TimeUnit::Years, 5.47),
    (12, TimeUnit::Years, 5.60),
    (15, TimeUnit::Years, 5.75),
    (20, TimeUnit::Years, 5.89),
    (25, TimeUnit::Years, 5.95),
    (30, TimeUnit::Years, 5.96),
];

fn main() {
    // --- 1. Settings (D5: explicit, no global singleton) ---------------------
    // Evaluation date MUST be set before helpers are built; under D5 an unset
    // date is an Err, not a clock fallback.
    let calendar = Target::new();
    let today = calendar.adjust(
        Date::new(15, Month::June, 2026),
        BusinessDayConvention::Following,
    );
    let settings = shared(Settings::<Date>::new());
    settings.set_evaluation_date(today);

    // Spot / settlement date = today + 2 business days; this is the curve's
    // reference date.
    let settlement = calendar.advance(
        today,
        2,
        TimeUnit::Days,
        BusinessDayConvention::Following,
        false,
    );

    // --- 2. Rate helpers -----------------------------------------------------
    // Each helper's index takes an EMPTY forwarding handle: the curve links
    // itself into the helpers during the bootstrap (passing the curve handle
    // here would create the circular dependency).
    let mut instruments: Vec<Shared<dyn RateHelper>> = Vec::new();

    // Deposit helpers. Euribor::new rejects TimeUnit::Days tenors, so the
    // shortest deposit here is 1W.
    for (n, units, rate) in DEPOSIT_DATA {
        let quote = Handle::new(shared(SimpleQuote::new(rate / 100.0)) as Shared<dyn Quote>);
        let index = Euribor::new(Period::new(n, units), Handle::empty(), settings.clone())
            .expect("deposit tenor is valid");
        instruments.push(DepositRateHelper::new(quote, &index) as Shared<dyn RateHelper>);
    }

    // Swap helpers: annual 30/360 fixed leg vs 6M Euribor float leg.
    for (n, units, rate) in SWAP_DATA {
        let quote = Handle::new(shared(SimpleQuote::new(rate / 100.0)) as Shared<dyn Quote>);
        let euribor6m = Euribor::six_months(Handle::empty(), settings.clone());
        instruments.push(SwapRateHelper::new(
            quote,
            Period::new(n, units),
            calendar.clone(),
            Frequency::Annual,
            BusinessDayConvention::Unadjusted,
            Thirty360::with_convention(Convention::BondBasis),
            &euribor6m,
        ) as Shared<dyn RateHelper>);
    }

    // --- 3. Build the curve --------------------------------------------------
    // <Discount, LogLinear>: node values are discount factors, log-linearly
    // interpolated. Construction is cheap; the bootstrap runs lazily on the
    // first read (discount / max_date / dates).
    let curve = PiecewiseYieldCurve::<Discount, LogLinear>::new(
        settlement,
        instruments,
        Actual360::new(),
        LogLinear,
    )
    .expect("curve constructed");

    // Forcing the bootstrap (and surfacing any bootstrap error) up front.
    let nodes = curve.dates().expect("bootstrap succeeds");
    println!("bootstrapped {} curve nodes", nodes.len());

    // --- 4. Query discount factors and zero rates ----------------------------
    // Pick a couple of forward dates inside the curve's range (extrapolate =
    // false): the 5Y and 10Y swap pillars.
    for years in [5, 10] {
        let date = settlement + Period::new(years, TimeUnit::Years);

        // discount_date(date): DF from the reference date to `date`.
        let df = curve
            .discount_date(date, false)
            .expect("date within curve range");

        // discount(t): same, but by time measured in the curve's own day count.
        let t = curve.time_from_reference(date).expect("time");
        let df_t = curve.discount(t, false).expect("t within range");

        // zero_rate(t): continuously-compounded zero rate for time t.
        let zero: Rate = curve
            .zero_rate(t, Compounding::Continuous, Frequency::Annual, false)
            .expect("zero rate")
            .rate();

        println!(
            "{years:>2}Y  t={t:.4}  DF(date)={df:.8}  DF(t)={df_t:.8}  zero(cont)={:.6}%",
            zero * 100.0
        );
    }
}
